Support pause deployments

0xmichalis opened the pull request:
This PR adds support for inert deployments so a user can choose
when to run a deployment that exists in the system instead of having
the deployment controller automatically reconciling it after every
change or sync interval.

@ironcladlou @bgrant0607 @janetkuo @nikhiljindal @smarterclayton @pweil- 

Closes https://github.com/kubernetes/kubernetes/issues/14516

ironcladlou commented:
The tri-state boolean seems awkward to me. It makes me feel like we're missing something in the design.

Another idea: what if inert was a primitive boolean? To override the inert state temporarily, the DeploymentRequest endpoint could set an annotation. The controller would then reconcile if not inert or if the annotation exists. Upon completed reconciliation the controller could then clear the annotation, leaving the deployment in its previous state.

kargakis commented:
@ironcladlou cool, I think what you propose is cleaner. I will update accordingly.

kargakis commented:
Somehow the isInert field gets swallowed without getting any validation error. Is there anything I need to update for the api changes? @nikhiljindal @bgrant0607 ?

nikhiljindal commented:
What is the use case for reconcile it once?
It seems to me that we are leaking internal implementation detail for reconcileDeployment. User should not depend on what reconcileDeployment does when its run once.

I see that the swagger spec is not updated. You need to run `hack/update-swagger-spec.sh`

nikhiljindal commented:
Unless there is a strong use case, we do not need the instantiate endpoint, DeploymentRequest and the annotation. Just the IsInert field is enough. DeploymentController will ignore the deployment if that is true. We can maybe provide a subresource to make changing IsInert easier.

kargakis commented:
> What is the use case for reconcile it once?

https://github.com/kubernetes/kubernetes/issues/14516#issuecomment-143037423

cc: @smarterclayton

kargakis commented:
> I see that the swagger spec is not updated. You need to run hack/update-swagger-spec.sh

It runs as part of the update-all script. Doesn't produce any additional diff when I'm running it separately.

ironcladlou commented on test/e2e/deployment.go:
```diff
@@ -288,3 +291,58 @@ func testRollingUpdateDeploymentEvents(f *Framework) {
+	req := &extensions.DeploymentRequest{Name: deploymentName}
+	deployment, err = c.Deployments(ns).Instantiate(req)
+	Expect(err).NotTo(HaveOccurred())
+	rcs, err = deploymentutil.GetRCsForDeployment(c, *deployment)
```
Will this race with the controller creating the RC?

kargakis replied:
Higlhy possible. I need to rework this part.

kargakis commented:
> It runs as part of the update-all script. Doesn't produce any additional diff when I'm running it separately.

Ok, found the problem. There are manual conversions for Deployments that I had to update:)

ironcladlou commented:
This lgtm now, but we need @nikhiljindal to weigh in on the latest design.

nikhiljindal commented:
It is still not clear to me whats the use case for "run syncDeployment once". What does "run once" mean to the user.

Controller should sync deployment if isInert is false, and should not sync it otherwise.

ironcladlou commented:
Talked with @bgrant0607 about this today offline and he suggested `Pause` as an alternative to `IsInert`. I like `Pause` a lot better. At this point I'm actually not entirely clear on the need for `/instantiate`. @smarterclayton want to defend the feature?

ironcladlou commented:
I'd rather scrap the `/instantiate` support and get the core pause concept in place. We can revisit later.

bgrant0607 commented:
Declarative: `Paused`. Similar to `Unschedulable`.

bgrant0607 commented:
Is `/instantiate` an alternative to hooks?

kargakis commented:
I replaced `IsInert` with `Paused`, removed the `/instantiate` endpoint and all its machinery, and re-run `update-all`. 

> Is /instantiate an alternative to hooks?

No, it is supposed to enable entirely manually triggered deployments. Toggling on and off `.spec.paused` is still manual but doesn't let the users know how many times their deployments are proceessed in between. By issuing a ` PUT /instantiate` I know that my deployment will be processed only once. We can come back to this later if we want it.

kargakis commented:
> Toggling on and off .spec.paused is still manual but doesn't let the users know how many times their deployments are proceessed in between.

Hm, actually this is not true. I guess I like the latest state of the PR more after realizing it.

nikhiljindal commented on pkg/registry/deployment/etcd/etcd.go:
```diff
@@ -136,9 +135,6 @@ func (r *ScaleREST) Get(ctx api.Context, name string) (runtime.Object, error) {
 }
 
 func (r *ScaleREST) Update(ctx api.Context, obj runtime.Object) (runtime.Object, bool, error) {
-	if obj == nil {
```
Is this deletion intended?

kargakis replied:
Yes, needless check. I will remove these file changes though.

nikhiljindal commented on test/e2e/deployment.go:
```diff
@@ -317,3 +322,71 @@ func testRolloverDeployment(f *Framework) {
+	case <-w.ResultChan():
+		// this is it
+	case <-time.After(time.Minute):
+		err = fmt.Errorf("watch timed out")
```
Would be better to add a more meaningful error message like "expected a new rc to be created"

kargakis replied:
ok

nikhiljindal commented on pkg/util/deployment/deployment.go:
```diff
@@ -123,6 +123,14 @@ func GetNewRCTemplate(deployment extensions.Deployment) api.PodTemplateSpec {
 }
 
+// GetRCsForDeployment returns all the replication controllers for the provided deployment.
+func GetRCsForDeployment(c client.Interface, deployment extensions.Deployment) ([]api.ReplicationController, error) {
```
I dont think that this function does what you expect it to do.
You want the RCs that have the same selector as deployment.Spec.Selector, not the ones selected by it.

For ex: it will fail for the following RCs and deployment (it will not return the RC it should).

Deployment {
 "spec" : {
     "selector": {
         "pod": "true",
      }
  }
}

RC {
  "labels": {
       "rc": "true",
   }
   "selector": {
     "pod": "true",
  }
}

You can use GetOldRCs and GetNewRCs instead.

kargakis replied:
Right, but since the rc selector (which is the same as the deployment selector) is propagated in the rc labels via defaulting, is this really a problem? Users shouldn't mess with rcs owned by deployments anyway.

Relevant issue:  https://github.com/kubernetes/kubernetes/issues/17491

nikhiljindal replied:
deployment can also adopt an existing RC for which this may or may not be true

kargakis replied:
ok, removed `GetRCsForDeployment`

nikhiljindal commented on test/e2e/deployment.go:
```diff
@@ -317,3 +322,71 @@ func testRolloverDeployment(f *Framework) {
+	Expect(err).NotTo(HaveOccurred())
+
+	// Verify that there are no realized states for the paused deployment.
+	rcs, err := deploymentutil.GetRCsForDeployment(c, *deployment)
```
You can check for RCs.
If its easier, I am fine with checking that there are no events on this deployment instead.
Your call.

kargakis replied:
I prefer checking for the actual rc rather than an event. Let's get the way we get the rc sorted out (GetRCsForDeployment vs GetOldRCs/GetNewRC.

nikhiljindal commented:
This looks a lot better now without `/instantiate`.
Added a few comments.

kargakis commented:
Most comment addressed. Still waiting @nikhiljindal on deciding the way to get the rcs for a deployment. I can change to the already existing functions but eventually we should optimize the way we do lists in the controller (https://github.com/kubernetes/kubernetes/issues/17491).

janetkuo commented on test/e2e/deployment.go:
```diff
@@ -317,3 +322,71 @@ func testRolloverDeployment(f *Framework) {
+	deploymentName := "nginx"
+	podLabels := map[string]string{"name": "nginx"}
+	Logf("Creating paused deployment %s", deploymentName)
+	_, err := c.Deployments(ns).Create(&extensions.Deployment{
```
Can you also add a test for creating a deployment that's not paused, and then update it to be paused, and the deployment should be paused after the update.

kargakis replied:
done

kargakis commented:
Squashed down to one commit

alex-mohr commented:
No CLA label, so closing and reopening to trigger.

0xmichalis committed:
extensions: support paused deployments

This commit adds support for paused deployments so a user can choose
when to run a deployment that exists in the system instead of having
the deployment controller automatically reconciling it after every
change or sync interval.

bgrant0607 commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -213,6 +213,10 @@ type DeploymentSpec struct {
 	// No label is added if this is set to empty string.
 	UniqueLabelKey *string `json:"uniqueLabelKey,omitempty"`
+
+	// Paused indicates that the deployment will not be processed by the deployment
```
Please either reference the json field name (not capitalized) or don't mention the field name.

Does Paused also block scaling behavior? I think it shouldn't.

bgrant0607 replied:
What's our current resize policy? Resize the largest RC (up or down)? If so, that seems ok, even while paused.

kargakis replied:
Well, client-side scaling intentionally updates only the deployment to avoid any races (ie the only entrypooint for scaling the active RC is the deployment controller) so yes, paused blocks scaling behavior since the deployment won't be reconciled.

kargakis replied:
Note that the deployment itself will be scaled so once it resumes, the controller will reconcile it, see that it needs to scale the active RC and will do so.

kargakis replied:
Updated the comment to not mention the field name and regenerated the swagger spec

bgrant0607 replied:
We'll want to figure out how to retain scaling while paused, but I'm ok with this during Beta.

bgrant0607 replied:
Filed #20130

kargakis commented:
test flake
```
!!! [0122 17:24:58] Timed out waiting for apiserver:  to answer at http://127.0.0.1:8080/healthz; tried 120 waiting 1 between each
!!! Error in ./hack/test-update-storage-objects.sh:48
  'return 1' exited with status 1
Call stack:
  1: ./hack/test-update-storage-objects.sh:48 startApiServer(...)
  2: ./hack/test-update-storage-objects.sh:123 main(...)
Exiting with status 1
```
@k8s-bot unit test this

kargakis commented:
@k8s-bot unit test this

nikhiljindal commented on test/e2e/deployment.go:
```diff
@@ -387,6 +392,98 @@ func testRolloverDeployment(f *Framework) {
+	podLabels := map[string]string{"name": "nginx"}
+	Logf("Creating paused deployment %s", deploymentName)
+	_, err := c.Deployments(ns).Create(&extensions.Deployment{
+		ObjectMeta: api.ObjectMeta{
```
Use newDeployment()

kargakis replied:
done

nikhiljindal commented:
LGTM, modulo one comment

kargakis commented:
all comments addressed; let's get this in so I can deliver rollout pause/resume w/o having to rebase every now and then both PRs:)

nikhiljindal commented:
LGTM, thanks!

kargakis commented:
@nikhiljindal also needs a cla tag:)

kargakis commented:
@bgrant0607 reverted all the package renames

this needs the lgtm label back

0xmichalis committed:
extensions: generated code for paused deployments

kargakis commented:
rebased

kargakis commented:
```
!!! Error in ./hack/test-cmd.sh:203
  '[ "$(kubectl get nodes -o go-template='{{ .apiVersion }}' "${kube_flags[@]}")" == "v1" ]' exited with status 1
Call stack:
  1: ./hack/test-cmd.sh:203 runTests(...)
  2: ./hack/test-cmd.sh:1352 main(...)
Exiting with status 1
```
test flake

kargakis commented:
@k8s-bot unit test this

bgrant0607 commented:
LGTM.

We're trying to eliminate test flakes, so please don't restart the unit/integration tests so that we can look at the logs (or you can help debug :-)).

kargakis commented:
> We're trying to eliminate test flakes, so please don't restart the unit/integration tests so that we can look at the logs (or you can help debug :-)).

I post every test output (the part that denotes the failure) before restarting the tests. So far I have stumbled upon flakes only

bgrant0607 commented:
@k8s-bot unit test this

bgrant0607 commented on pkg/storage/testing/types.generated.go.tmp:
```diff
@@ -0,0 +1,365 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
```
This might be the cause of the test failure and, anyway, we shouldn't commit these tmp files. Please remove it from the PR.

kargakis replied:
doh

done

kargakis commented:
@bgrant0607 tests are green now

bgrant0607 commented:
LGTM ,thanks

kargakis commented:
```
!!! [0127 00:41:46] Timed out waiting for apiserver:  to answer at http://127.0.0.1:8080/healthz; tried 120 waiting 1 between each
!!! Error in ./hack/test-update-storage-objects.sh:48
  'return 1' exited with status 1
Call stack:
  1: ./hack/test-update-storage-objects.sh:48 startApiServer(...)
  2: ./hack/test-update-storage-objects.sh:123 main(...)
Exiting with status 1
```
Seems like an apiserver timeout flake. Rekicking tests...

kargakis commented:
@k8s-bot unit test this
